### ISSUE TITLE
REST API app can have all requests filtered out

### ISSUE THREAD

**Issue by SergeyKanzhelev** [Microsoft/ApplicationInsights-dotnet-server#175]
Reported by the customer:

> What I notice when running locally, the request fails it contains the response code but when its successful then it doesn’t I am not sure why?
> 
> This is RESTApi Project created using Visual Studio template for Azure Cloud Service -> WebRole -> Empty (Web API Check box) targeting .NET 4.5.2. Project just contain REST Endpoints there is no MVC used. Application Insights configured using Visual Studio (right click on the project and select add AI). 

Removing filters from the collection helped to resolve the issue:

``` xml
<Add Type="Microsoft.ApplicationInsights.Web.RequestTrackingTelemetryModule, Microsoft.AI.Web">
  <Handlers>
<!-- 
        Add entries here to filter out additional handlers: 

        NOTE: handler configuration will be lost upon NuGet upgrade.

     Remove handlers below:
-->
     <Add>System.Web.Handlers.TransferRequestHandler</Add>
     <Add>Microsoft.VisualStudio.Web.PageInspector.Runtime.Tracing.RequestDataHttpHandler</Add>
     <Add>System.Web.StaticFileHandler</Add>
     <Add>System.Web.Handlers.AssemblyResourceLoader</Add>
     <Add>System.Web.Optimization.BundleHandler</Add>
     <Add>System.Web.Script.Services.ScriptHandlerFactory</Add>
     <Add>System.Web.Handlers.TraceHandler</Add>
     <Add>System.Web.Services.Discovery.DiscoveryRequestHandler</Add>
     <Add>System.Web.HttpDebugHandler</Add>
    </Handlers>
</Add>
```

**Comment by SergeyKanzhelev** [comment]
BTW, this issue may be a continuation of this one: https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/16

**Comment by codecadwallader** [comment]
I was having the same issue, and also saw that commenting out the one line in the ApplicationInsights.config caused successful requests to start getting logged.  Web API 2 project.

```xml
        <!--<Add>System.Web.Handlers.TransferRequestHandler</Add>-->
```

**Comment by SergeyKanzhelev** [comment]
@codecadwallader thanks for reporting!

**Comment by cijothomas** [comment]
Not reproed. With the default AI Config I am getting all (sucess,failed) requests in WebApi project.

**Comment by cijothomas** [comment]
Tried both empty ASP.NET with WebAPI, and AzureCloudService project - both working fine.

**Comment by SergeyKanzhelev** [comment]
Couple ideas:
- Maybe try running an app in IIS Here is how `TransferRequestHandler` [used](https://blogs.msdn.microsoft.com/tmarq/2010/05/26/how-extensionless-urls-are-handled-by-asp-net-v4/)
- I believe we are filtering it so WCF requests wouldn't be reported twice. You can try running functional tests without it to confirm.

**Comment by cijothomas** [comment]
I believe we are filtering it so WCF requests wouldn't be reported twice. You can try running functional tests without it to confirm. - This is true. The WCF functional test fail as it gets 2 items instead of 1, if TransferRequestHandler  is commented out.

**Comment by SergeyKanzhelev** [comment]
Suggested fix from @drago-draganov:

A possible solution is to use HttpRequest.Headers. Request Headers are carries from parent to child request. The application can add custom request headers (ex: AppInsights-Request-Id).

**HttpModule**

``` csharp
const string REQUEST_ID_HEADER = “AppInsights-Request-Id”;
static ConcurrentDictionary<string, string> _activeRequests = new ConcurrentDictionary<string, string>();
```

**BeginRequest:**

```
if (Request.Headers[REQUEST_ID_HEADER] == null) {
       string id = Guid.NewGuid().ToString();

       _activeRequests.TryAdd(id, id);
        Request.Headers[REQUEST_ID_HEADER] = id;
} 
```

**EndRequest:**

``` csharp
string requestId;
If (_activeRequests.TryRemove(REQUEST_ID_HEADER, out requestId)) {
                //
                // Log request data here
                //
}
```

The child request will remove the id from `_activeRequests` at EndRequest, so the parent won’t find it later, therefore missing the log. If there isn’t child request, at EndRequest the id will be in `_activeRequests` and perform log.

Begin of the child will happen before the End of the parent. Child requests always belongs to the same appdomain.

**Comment by cijothomas** [comment]
Increased severity, as more people have reported being affected with this issue.

**Comment by RPM1984** [comment]
FWIW - we just ran into this issue for all our production OWIN-based (Nancy) web apps (with no code changes/deployments). Commenting out the `TransferRequestHandler` key worked for us also.

**Comment by Bio2hazard** [comment]
We use OWIN (Katana), and I noticed that incoming requests get flagged as TransferRequest after passing through the Microsoft.Owin.Security middleware ( specifically after `app.UseOAuthAuthorizationServer(oAuthServerOptions);` ). 

Since WebApi has to come after OAuth, it means that by default App Insights will only capture requests going to the security middleware.

Commenting out the TransferRequestHandler was necessary to capture requests hitting the WebApi middleware.

**Comment by benburton-se** [comment]
I also have this issue in my Web API. After much searching I've finally found the answer here. Removing the TransferRequestHandler from ApplicationInsights.config worked for me. I'm not sure I can provide much more on specifically which circumstances this problem occurs under though

**Comment by cijothomas** [comment]
@benburton-se The fix for this issue is in the works already and is planned for next release. beta2.
PR here: https://github.com/Microsoft/ApplicationInsights-dotnet-server/pull/699

**Comment by MS-TimothyMothra** [comment]
the fix for this issue has been merged and will be released as part of 2.5 beta 2
